agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/ClientDecorator.java

```diff
@@ -15,10 +15,51 @@
  */
 package io.opentelemetry.auto.bootstrap.instrumentation.decorator;
 
+import io.grpc.Context;
+import io.opentelemetry.context.ContextUtils;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.DefaultSpan;
 import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.Tracer;
+import io.opentelemetry.trace.TracingContextUtils;
 
 public abstract class ClientDecorator extends BaseDecorator {
 
+  // Keeps track of the client span in a subtree corresponding to a client request.
+  private static final Context.Key<Span> CONTEXT_CLIENT_SPAN_KEY =
+      Context.key("opentelemetry-trace-auto-client-span-key");
+
+  public static Scope currentContextWith(final Span clientSpan) {
+    if (!clientSpan.getContext().isValid()) {
+      return TracingContextUtils.currentContextWith(clientSpan);
+    }
+    return ContextUtils.withScopedContext(
+        TracingContextUtils.withSpan(
+            clientSpan, Context.current().withValue(CONTEXT_CLIENT_SPAN_KEY, clientSpan)));
+  }
+
+  public static Context withSpan(final Span clientSpan, final Context context) {
+    if (!clientSpan.getContext().isValid()) {
+      return TracingContextUtils.withSpan(clientSpan, context);
+    }
+    return TracingContextUtils.withSpan(
+        clientSpan, Context.current().withValue(CONTEXT_CLIENT_SPAN_KEY, clientSpan));
+  }
+
+  public Span getOrCreateSpan(String name, Tracer tracer) {
+    final Context context = Context.current();
+    final Span clientSpan = CONTEXT_CLIENT_SPAN_KEY.get(context);
+
+    if (clientSpan != null) {
+      // We don't want to create two client spans for a given client call, suppress inner spans.
+      return DefaultSpan.getInvalid();
+    }
+
+    final Span current = TracingContextUtils.getSpan(context);
+    return tracer.spanBuilder(name).setSpanKind(Kind.CLIENT).setParent(current).startSpan();
```

Comment:
Here we are sure that we are currently creating first client span. I think this should be the right place to put that span into current context under `CONTEXT_CLIENT_SPAN_KEY` key.

Reply:
I really wish I could - but context is immutable so AFAIK, we can really only "add to context" when mounting it, in these `withSpan` type of methods. I have left some comments in the context discussion doc before coincidentally, and it is running into these sort of cases where I really don't enjoy immutable context ;)

If there is an entrypoint that is guaranteed to be called with a new context, than it would be possible to put mutable state in at that point and then here we'd just update it. But I couldn't find one, if this span is the root of a trace (some non-server app running AWS SDK call for example), this span will be created before a new context is and we wouldn't have anywhere to store it until a call to `with*`.

Let me know if I'm missing anything, I'd be happy to be since this API is otherwise somewhat ackward for duplicating the APIs in `TracingContextUtils`.

Reply:
I see what you mean. As long as you return `Span` from this method, you really cannot modify `Context`